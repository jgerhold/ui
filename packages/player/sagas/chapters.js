import { propOr, prop } from 'ramda'
import { takeEvery, select, put } from 'redux-saga/effects'
import {
  INIT,
  REQUEST_PLAYTIME,
  BACKEND_PLAYTIME,
  SET_CHAPTER,
  DISABLE_GHOST_MODE,
  PREVIOUS_CHAPTER,
  NEXT_CHAPTER
} from '@podlove/player-actions/types'
import * as config from '@podlove/utils/config'
import * as chapter from '@podlove/player-actions/chapters'
import { requestPlaytime } from '@podlove/player-actions/timepiece'
import { toPlayerTime } from '@podlove/utils/time'
import { hostname } from '@podlove/utils/url'

export const chaptersSaga = ({
  selectDuration,
  selectPlaytime,
  selectCurrentChapter,
  selectChapterList
}) =>
  function* saga() {
    yield takeEvery(INIT, initChapters, {
      selectDuration,
      selectPlaytime,
      selectCurrentChapter,
      selectChapterList
    })
    yield takeEvery(REQUEST_PLAYTIME, chapterUpdate)
    yield takeEvery(BACKEND_PLAYTIME, chapterUpdate)
    yield takeEvery(SET_CHAPTER, setChapter, {
      selectDuration,
      selectPlaytime,
      selectCurrentChapter,
      selectChapterList
    })
    yield takeEvery(DISABLE_GHOST_MODE, resetChapter, {
      selectDuration,
      selectPlaytime,
      selectCurrentChapter,
      selectChapterList
    })
    yield takeEvery(PREVIOUS_CHAPTER, previousChapter, {
      selectDuration,
      selectPlaytime,
      selectCurrentChapter,
      selectChapterList
    })
    yield takeEvery(NEXT_CHAPTER, nextChapter, {
      selectDuration,
      selectPlaytime,
      selectCurrentChapter,
      selectChapterList
    })
  }

export function* chapterUpdate({ payload }) {
  yield put(chapter.updateChapter(payload))
}

export function* setChapter({ selectCurrentChapter }) {
  const current = yield select(selectCurrentChapter)
  yield put(requestPlaytime(current.start))
}

export function* resetChapter({ selectPlaytime }) {
  const playtime = yield select(selectPlaytime)
  yield put(chapter.updateChapter(playtime))
}

export function* previousChapter({ selectPlaytime, selectCurrentChapter }) {
  const playtime = yield select(selectPlaytime)
  const { start, index } = yield select(selectCurrentChapter)

  if (playtime - start <= 1000) {
    yield put(chapter.previousChapter())
  } else {
    yield put(chapter.setChapter(index - 1))
  }
}

export function* nextChapter({
  selectDuration,
  selectPlaytime,
  selectChapterList,
  selectCurrentChapter
}) {
  const duration = yield select(selectDuration)
  const playtime = yield select(selectPlaytime)
  const chapters = yield select(selectChapterList)
  const { start, index } = yield select(selectCurrentChapter)

  const chapterStart = index === chapters.length && playtime >= start ? duration : start

  yield put(requestPlaytime(chapterStart))
}

export function* initChapters({ selectDuration }, { payload }) {
  const chapters = config.chapters(payload)
  const duration = yield select(selectDuration)

  const state = chapters.reduce((result, chapter, index, chapters) => {
    const end = propOr({ start: duration }, index + 1, chapters)
    const href = propOr(null, 'href', chapter)

    return [
      ...result,
      {
        index: index + 1,
        start: toPlayerTime(chapter.start),
        end: toPlayerTime(end.start),
        title: prop('title', chapter),
        image: prop('image', chapter),
        href,
        linkTitle: href ? hostname(href) : null
      }
    ]
  }, [])

  yield put(chapter.setChapters(state))
}
