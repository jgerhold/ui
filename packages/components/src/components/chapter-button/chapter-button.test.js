import { mount } from '@vue/test-utils'
import { nextChapter, previousChapter } from '@podlove/player-actions/chapters'

import ChapterButton from './ChapterButton'

describe('ChapterButton', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(ChapterButton, {
        propsData: {
          type: 'next'
        }
      })

      expect(wrapper.isVueInstance()).toBeTruthy()
    })
  })

  describe('props', () => {
    const types = ['next', 'previous']

    types.forEach(type => {
      test(`should render type '${type}' with prop color`, async () => {
        const wrapper = await mount(ChapterButton, {
          propsData: {
            color: '#fff',
            type
          }
        })

        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('events', () => {
    test('should emit NEXT_CHAPTER on next click', async () => {
      const wrapper = await mount(ChapterButton, {
        propsData: {
          type: 'next'
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted('click')).toEqual([[nextChapter()]])
    })

    test('should emit PREVIOUS_CHAPTER on next click', async () => {
      const wrapper = await mount(ChapterButton, {
        propsData: {
          type: 'previous'
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted('click')).toEqual([[previousChapter()]])
    })
  })
})
