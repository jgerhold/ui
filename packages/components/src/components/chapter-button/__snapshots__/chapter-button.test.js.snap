// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChapterButton props should render type 'next' with prop color 1`] = `
<button
  class="chapter-button"
  id="chapter-button--next"
>
  <svg
    aria-hidden="true"
    background="currentColor"
    height="15"
    version="1.1"
    viewBox="0 0 14 15"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <polygon
      fill="#fff"
      fill-rule="evenodd"
      points="0 0.4245 0 14.9995 12.6225 7.712"
      stroke="none"
    />
     
    <polygon
      fill="#fff"
      fill-rule="evenodd"
      points="10.435 15 13.87125 15 13.87125 0.425 10.435 0.425"
      stroke="none"
    />
  </svg>
   
</button>
`;

exports[`ChapterButton props should render type 'previous' with prop color 1`] = `
<button
  class="chapter-button"
  id="chapter-button--previous"
>
  <svg
    aria-hidden="true"
    background="currentColor"
    height="15"
    version="1.1"
    viewBox="0 0 14 15"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <polygon
      fill="#fff"
      fill-rule="evenodd"
      points="13.872375 0.4245 1.249875 7.712 13.872375 14.9995"
      stroke="none"
    />
     
    <polygon
      fill="#fff"
      fill-rule="evenodd"
      points="0 15 3.4375 15 3.4375 0.425 0 0.425"
      stroke="none"
    />
  </svg>
   
</button>
`;
